### ISSUE TITLE
[MNG-6261] Relative parent POM resolution failing in 3.5.0 with complex multimodule builds

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7607]
**[Dawid Weiss](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dweiss)** opened **[MNG-6261](https://issues.apache.org/jira/browse/MNG-6261?redirect=false)** and commented

In my effort to upgrade an existing (fairly complex) Maven build to Java 1.9.0 I updated Maven to 3.5.0 (from 3.3.9). Unfortunately I get errors when the project's modules are resolved:

```
> mvn validate
[FATAL] Non-resolvable parent POM for com.carrotsearch.lingo4g:lingo4g-public-bom:[unknown-version]: Could not find artifact com.carrotsearch.lingo4g:lingo4g-public:pom:1.6.0-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 5, column 11
...
(and many follow).
```

This build has a correct pom parent-structure (a tree), but is fairly complex -- the submodule hierarchy is not aligned with parent-child pom hierarchy (it's best to look at the repro code to understand how it's structured).

However, it's been working correctly with all prior Maven versions and I wonder if it's a regression bug or maybe underspecified Maven requirement (that should be enforced with a warning and not lead to this odd runtime message).

---

**Affects:** 3.5.0

**Attachments:**
- [capture-6.png](https://issues.apache.org/jira/secure/attachment/12879167/capture-6.png) (_134.41 kB_)
- [repro.zip](https://issues.apache.org/jira/secure/attachment/12877275/repro.zip) (_2.33 kB_)

**Issue Links:**
- [MNG-6503](https://issues.apache.org/jira/browse/MNG-6503) parent pom relative path regression after 3.3.9
 (_**"is duplicated by"**_)
- [MNG-6583](https://issues.apache.org/jira/browse/MNG-6583) Relative path for parent pom on Windows fails depending on case of drive letter
 (_**"is duplicated by"**_)
- [MNG-6262](https://issues.apache.org/jira/browse/MNG-6262) getCanonicalFile() is not used consistently during project resolution

**Remote Links:**
- [GitHub Pull Request #192
](https://github.com/apache/maven/pull/192)

1 votes, 8 watchers

**Comment by jira-importer** [comment]
**[Dawid Weiss](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dweiss)** commented

An additional element of difficulty: the fail behavior is not deterministic with respect to where the project resides in the path. I have the same (identical) git repo in two different locations (drives on Windows) and one passes with 3.5.0, while the other fails.

**Comment by jira-importer** [comment]
**[Dawid Weiss](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dweiss)** commented

And I now know where the problem is. Try this, Robert: open up cmd and switch to another drive:

```
> d:
```

Note the drive's path should be case-sensitive (so lowercase in this example). I don't know how or why Windows does it, but if you type (the first time):

```
> D:
```

it'll be upper-case cwd. Now, try to run 'mvn validate' on the project I attached as the repro. In one of these windows the build will succeed, in the other it'll fail.

Locating the cause of the problem was fairly easy from there on. I noticed that the failing build attempts to resolve certain POMs that should have been in the build's reactor:

```
[DEBUG] Could not find metadata com.carrotsearch.lingo4g:lingo4g-public:1.6.0-SNAPSHOT/maven-metadata.xml in local (E:\bamboo-home-3\.m2\AGENT-2752517\repository)
```

this was odd (and not present in the successfull build). So I dug down and dumped some stack traces to see how this happens. Turns out the problem is this fragment in `DefaultModelBuilder`:

```
                /*
                 * NOTE: This is a sanity check of the cache hit. If the cached parent POM was locally resolved, the
                 * child's <relativePath> should point at that parent, too. If it doesn't, we ignore the cache and
                 * resolve externally, to mimic the behavior if the cache didn't exist in the first place. Otherwise,
                 * the cache would obscure a bad POM.
                 */

                File pomFile = parentData.getModel().getPomFile();
                if ( pomFile != null )
                {
                    ModelSource expectedParentSource = getParentPomFile( childModel, childSource );

                    if ( expectedParentSource instanceof ModelSource2
                        && !pomFile.toURI().equals( ( (ModelSource2) expectedParentSource ).getLocationURI() ) )
                    {
                        parentData = readParentExternally( childModel, request, problems );
                    }
                }
```

Looks innocent, but those location URIs on Windows give me a false positive on the drive letter, for example:

```
file:/d:/_tmp/repro/lingo4g-public/pom.xml 
file:/D:/_tmp/repro/lingo4g-public/pom.xml
```

One of these is relative and is resolved using `getAbsoluteFile` which (I believe, didn't check) uses an upper-case drive letter. The other is resolved based on the cwd.

The ideal solution and fix for this would be to use NIO2's utilities (`Files.isSameFile` for comparing `Path` instances. The filesystem provider knows whether the filesystem is case sensitive or not. But even with `File` instances resolving canonical File instances and using `File.compareTo(File)` would be a better solution than comparing URIs.

A patch here is trivial (although exact solution is probably dependent on the variant chosen, as I mentioned), so I don't include a patch.

**Comment by jira-importer** [comment]
**[Dawid Weiss](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dweiss)** commented

One more comment - to get the 'different' drive letter on windows you have to sweat a bit (use `/d` switch to enforce full path change):

```
1. switch to c drive (or a different drive)
> c:
1. switch to lowercase d drive:
> cd /d d:\
1. switch to uppercase D drive:
> cd /d D:\
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

The entire file handling needs to be changed for `Path`.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Comparisson is done in org.apache.maven.model.building.DefaultModelBuilder.readParent(), and the following code now returns `false` when there's a difference between the drive letter:

```
else
            {
                /*
                 * NOTE: This is a sanity check of the cache hit. If the cached parent POM was locally resolved, the
                 * child's <relativePath> should point at that parent, too. If it doesn't, we ignore the cache and
                 * resolve externally, to mimic the behavior if the cache didn't exist in the first place. Otherwise,
                 * the cache would obscure a bad POM.
                 */

                File pomFile = parentData.getModel().getPomFile();
                if ( pomFile != null )
                {
                    ModelSource expectedParentSource = getParentPomFile( childModel, childSource );

                    if ( expectedParentSource instanceof ModelSource2
                        && !pomFile.toURI().equals( ( (ModelSource2) expectedParentSource ).getLocationURI() ) )
                    {
                        parentData = readParentExternally( childModel, request, problems );
                    }
                }
            }
```

**Comment by jira-importer** [comment]
**[Dawid Weiss](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dweiss)** commented

Thanks Robert.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I think the cleanest fix is to wrap pomFile in a `FileModelSource` and do an `equals()` with `expectedParentSource`. This also means that it is easy to unittest , verify all combinations of drive letters with FileModelSource.equals(). Hence I marked it as for-the-grabs. Now waiting for somebody to pick this up.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Almost there. In this case the unittest should be good enough. An explicit IT could be nice but is not necessary here.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [e4e33f73b8f089e443a3bcc678f509bf0daffb15](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=e4e33f73b8f089e443a3bcc678f509bf0daffb15)
Thanks for the patch!

**Comment by jira-importer** [comment]
**[Fabiano Cipriano de Oliveira](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fabcipriano)** commented

Thank you all for your support
